axum-macros/src/debug_handler.rs

```diff
@@ -284,6 +286,169 @@ fn check_inputs_impls_from_request(item_fn: &ItemFn, state_ty: Type) -> TokenStr
+fn check_output_tuples(item_fn: &ItemFn) -> Option<TokenStream> {
+    //Extract tuple types
+    let elements = match &item_fn.sig.output {
+        ReturnType::Type(_, ty) => match &**ty {
+            Type::Tuple(tuple) => &tuple.elems,
+            _ => return None,
+        },
+        _ => return None,
+    };
+
+    if elements.len() < 2 {
+        return None;
+    }
+    //Amount of IntoRequestParts
+    let mut parts_amount: i8 = 0;
+
+    let token_stream = WithPosition::new(elements.iter())
+        .enumerate()
+        .map(|(_idx, arg)| match &arg {
+            //First element type in the tuple
+            Position::First(ty) => {
+                let typename = extract_clean_typename(ty);
+                if typename.is_none() {
+                    quote! {}
+                } else {
+                    let typename = typename.unwrap();
+                    match &*typename.to_string() {
+                        "Parts" => quote! {},
+                        "Response" => quote! {},
+                        "StatusCode" => {
+                            quote! {}
+                        }
+                        _ => {
+                            parts_amount += 1;
+                            check_into_response_parts(ty, parts_amount)
+                        }
+                    }
+                }
+            }
+            Position::Last(ty) => check_into_response(ty),
+            Position::Middle(ty) => {
+                parts_amount += 1;
+                if parts_amount > 16 {
+                    let error_message = format!("Output Tuple cannot have more than 16 arguments.");
+                    let error = syn::Error::new_spanned(&item_fn.sig.output, error_message)
+                        .to_compile_error();
+                    error
+                } else {
+                    let name = _check_into_response_not_parts(ty);
+                    match name {
+                        Some(_) => {
+                            quote_spanned!{ty.span()=>
+                                compile_error!("This type only implements IntoResponse but not IntoResponseParts, try moving it to the last element");
+                            }
+                        },
+                        None => check_into_response_parts(ty, parts_amount)
+                    }
+                }
+            }
+            _ => quote! {},
+        })
+        .collect::<TokenStream>();
+    Some(token_stream)
+}
+
+fn _check_into_response_not_parts(ty: &Type) -> Option<&'static str> {
+    let name = extract_clean_typename(ty);
+    match name {
+        Some(name) => {
+            let s = match &*name {
+                "Json" => "Json<_>",
+                "String" => "String",
+                //todo add more types that implement IntoResponse but not IntoResponseParts
+                _ => return None,
+            };
+            Some(s)
+        }
+        None => None,
+    }
+}
```

Comment:
We already have the function https://github.com/tokio-rs/axum/blob/4cd8498aeb723ba7eda85a4eaf7680173c085b6f/axum-macros/src/debug_handler.rs#L531 which does something similar. Could we use that instead?

Reply:
Some of the types that that one handles don't implement `IntoResponse`, so I think we should make another helper function like that one, `response_body_provider_type_names` or whatever.

Reply:
Thanks for the review! I fixed all the problems except for this one, we may need a separate list for this one I think.

Reply:
> we may need a separate list for this one

Yes, that is what I was saying (unless I'm misunderstanding) 😅

Reply:
Yeah exactly, I forgot to tag your response😓.
I could test on some common used types and provide a more complete list for this if necessary.